### ISSUE TITLE
Forced Colors Mode

### ISSUE THREAD

**Issue by alisonmaher** [mozilla/standards-positions#463]
## Request for Mozilla Position on an Emerging Web Specification

* Specification Title: CSS Color Adjustment Module Level 1
* Specification or proposal URL: https://www.w3.org/TR/css-color-adjust-1/#forced
* Caniuse.com URL (optional): https://caniuse.com/mdn-css_at-rules_media_forced-colors
* Bugzilla URL (optional): https://bugzilla.mozilla.org/show_bug.cgi?id=1591210
https://bugzilla.mozilla.org/show_bug.cgi?id=1659511
https://bugzilla.mozilla.org/show_bug.cgi?id=1663200
* Mozillians who can provide input (optional): @emilio

### Other information
Given that a version of Forced Colors Mode has already shipped in Gecko, the main focus of this review is around the MQ and CSS aspects of the Forced Colors Mode feature.

**Comment by emilio** [comment]
cc @mreschenberg

**Comment by emilio** [comment]
So to be clear this is about:

 * https://drafts.csswg.org/css-color-adjust-1/#forced about forced-color-mode.
 * `@media (forced-colors)`: https://drafts.csswg.org/mediaqueries-5/#descdef-media-forced-colors
 * `@media (prefers-contrast: forced)`: https://drafts.csswg.org/mediaqueries-5/#prefers-contrast

Regarding the general feature, I think we're generally supportive. As you said Gecko is already shipping and has shipped for a long time something like this, based on computed-value-time adjustments / `revert` rather than used-value-time adjustments as the latest draft requires.

I think my main concerns about the used-value time adjustments are:

 * User stylesheets not getting respected if they don't use system colors (which sounds actually quite a bit annoying?).

 * That it depends on the resolution of https://github.com/w3c/csswg-drafts/issues/3847, but we left that with a TODO about interpolation. We then decided that `color-mix` was the tool for interpolation in https://github.com/w3c/csswg-drafts/issues/5392, but that doesn't seem implemented in Chromium, so I'm confused about what y'all do there. Do you know what does Chromium do when you get asked to interpolate one system color and another color? Do you eagerly resolve it? (that might be a reasonable solution in the interim)

Regarding the media queries, I think my main concern is https://github.com/w3c/csswg-drafts/issues/5433. I think that that duplication is quite unfortunate, but I don't think we reached a conclusion at TPAC.

Do you know what does Chromium plan to ship wrt the media queries? It'd be nice if we resolved on that issue before those media features get shipped on Chromium...

**Comment by emilio** [comment]
I filed https://github.com/w3c/csswg-drafts/issues/5779 on the above, and https://github.com/w3c/csswg-drafts/issues/5778 on something else I noticed while thinking about that :)

**Comment by alisonmaher** [comment]
Thanks Emilio! I appreciate the feedback. Regarding your questions and concerns:

> User stylesheets not getting respected if they don't use system colors (which sounds actually quite a bit annoying?).

Yeah that's a good point, although user stylesheets could also use `forced-color-adjust: none` to accomplish this when appropriate.

> That it depends on the resolution of w3c/csswg-drafts#3847, but we left that with a TODO about interpolation. We then decided that color-mix was the tool for interpolation in w3c/csswg-drafts#5392, but that doesn't seem implemented in Chromium, so I'm confused about what y'all do there. Do you know what does Chromium do when you get asked to interpolate one system color and another color? Do you eagerly resolve it? (that might be a reasonable solution in the interim)

@kbabbitt has been working on the implementation of system colors computing to themselves, so he can probably fill in some of the details that I miss here. However, it sounds like we are currently resolving the keywords eagerly during interpolation. Kevin is planning to open a CSSWG issue around this to clarify the expected behavior.

In any event, we had implemented the move of forced colors to used-value time such that it wasn't dependent on the system colors computing to themselves feature shipping. We are currently storing both the keyword and resolved value for system colors to accomplish this at used-value time while the system color work is still underway.

> Regarding the media queries, I think my main concern is w3c/csswg-drafts#5433. I think that that duplication is quite unfortunate, but I don't think we reached a conclusion at TPAC.
> 
> Do you know what does Chromium plan to ship wrt the media queries? It'd be nice if we resolved on that issue before those media features get shipped on Chromium...

As a part of shipping Forced Colors Mode, we are planning to ship the `forced-colors` media feature. Based on the issue you referenced, it was pretty clear that there was at least consensus on keeping that MQ. In regards to `prefers-contrast`, the plan is to ship that at a later point in time once that issue has been resolved.

**Comment by kbabbitt** [comment]
> > That it depends on the resolution of [w3c/csswg-drafts#3847](https://github.com/w3c/csswg-drafts/issues/3847), but we left that with a TODO about interpolation. We then decided that color-mix was the tool for interpolation in [w3c/csswg-drafts#5392](https://github.com/w3c/csswg-drafts/issues/5392), but that doesn't seem implemented in Chromium, so I'm confused about what y'all do there. Do you know what does Chromium do when you get asked to interpolate one system color and another color? Do you eagerly resolve it? (that might be a reasonable solution in the interim)
> 
> @kbabbitt has been working on the implementation of system colors computing to themselves, so he can probably fill in some of the details that I miss here. However, it sounds like we are currently resolving the keywords eagerly during interpolation. Kevin is planning to open a CSSWG issue around this to clarify the expected behavior.

Correct, currently the code eagerly resolves system colors to used values for interpolation. I had tossed out the idea of using color-mix() for the interpolated value in w3c/csswg-drafts#5392 but I don't know that it actually got agreed to. I opened https://github.com/w3c/csswg-drafts/issues/5780 to get this clarified.

**Comment by zcorpan** [comment]
The bugs are fixed. Per https://developer.mozilla.org/en-US/docs/Web/CSS/@media/forced-colors and https://developer.mozilla.org/en-US/docs/Web/CSS/forced-color-adjust it shipped in v89 and v113, respectively.

@emilio I suppose this should be `positive`?
